Functions/xASL_quant_HadamardDecoding.m

```diff
@@ -0,0 +1,165 @@
+function [imASL_reorder] = xASL_quant_HadamardDecoding(Encoded_ASL, xDecodingFields, NumberEchoTimes)
+%xASL_im_HadamardDecoding Hadamard-4 & Hadamard-8 Decoding
+%
+% FORMAT:       [Decoded_ASL] = xASL_im_HadamardDecoding(HadamardType, sec)
+%
+% INPUT:        HadamardType  - 4 for Hadamard4
+%                             - 8 for Hadamard8
+%                             - 12 for Hadamard12
+%               sec           - output from xASL_im_HadamardCaseByCase
+%
+%
+% -----------------------------------------------------------------------------------------------------------------------------------------------------
+% DESCRIPTION:  Hadamard-4 & Hadamard-8 Decoding.
+%
+% -----------------------------------------------------------------------------------------------------------------------------------------------------
+% EXAMPLE:      n/a
+% __________________________________
+% Copyright 2015-2021 ExploreASL
+
+% Decoding Fields
+
+TimeEncodedMatrixType = xDecodingFields.TimeEncodedMatrixType;
+TimeEncodedMatrixSize = xDecodingFields.TimeEncodedMatrixSize;
+
+if isfield(xDecodingFields,'DecodingMatrix') && ~isempty(xDecodingFields.DecodingMatrix)
+    DecodingMatrix_input = xDecodingFields.DecodingMatrix;
+else
+    DecodingMatrix_input = [];
+end
+
+% #### For Walsh Decoding Matrix ####
+if strcmp(TimeEncodedMatrixType,'Walsh')
+    
+    if TimeEncodedMatrixSize == 4
+        
+        ASL_im = xASL_io_Nifti2Im(Encoded_ASL); % Load time-series nifti
```

Comment:
Nice! Note that `xASL_io_Nifti2Im` is "transparent" in the sense `Encoded_ASL` here can be a path to a NIfTI file but can also already be a matrix; it will work in both cases. You can add this to the header.